### PR TITLE
C++: use struct tag instead of symbol_type

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 7128f2656ce25b76a3a57a16867dfd8276c28a7b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86346729
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 66c428b40b97cf3fef0a5e1dcc356bdeb46ad4d0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86348447

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 2165f2915174fd1cabec9f068813cdf7eeeab810).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86362376

File: src/cpp/cpp_typecheck_bases.cpp [hunk 1]
```diff
@@ -67,6 +70,8 @@ void cpp_typecheckt::typecheck_compound_bases(struct_typet &type)
       throw 0;
     }
 
+    base_symbol_expr.type().id(ID_symbol_type);
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_bases.cpp]: That seems redundant? I don't think there is any way we could get here with `base_symbol_expr.type().id() != ID_symbol_type`.

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a15fd5fe1914a1f081de0ca71d8f5a586e280a79).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86394591

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: eed524601ccdfe75c8fd128fa329a4a3fbcfa2a5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86707663

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 114e4aa7f57af009625dc262e768e28e512a1abd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86902083

File: src/cpp/cpp_typecheck_bases.cpp (hunk 1 above)
Comment (kroening) [reply src/cpp/cpp_typecheck_bases.cpp]: Indeed, turned into invariant.

File: src/cpp/cpp_typecheck_bases.cpp [hunk 2]
```diff
@@ -67,6 +70,10 @@ void cpp_typecheckt::typecheck_compound_bases(struct_typet &type)
       throw 0;
     }
 
+    DATA_INVARIANT(
+      base_symbol_expr.type().id() == ID_symbol_type,
+      "base classes are identified using symbol_types");
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_bases.cpp]: I'm still confused: in line 57 there is a `to_symbol_type`, and none of the code that's between lines 57 and this one touches `base_symbol_expr`.

Comment (kroening) [reply src/cpp/cpp_typecheck_bases.cpp]: Indeed, the invariant is unneeded, will remove.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8d06e52db07be6102fe6196015f19a865a7fa0ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87022823
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d2f803abb87f4c66c5dfcafe83b0ea9379775961).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87025201

**Pull request merged**